### ISSUE TITLE
Feature Request: tsd for nested delimiters

### ISSUE THREAD

**Issue by fuzzybear3965** [lervag/vimtex#258]
It would be cool if tsd could be extended to work over nested delimiters. Consider a case where an equation has many pairs of parentheses. I select the equation using visual mode. Then, I type ``tsd`` and all of the pairs of parentheses are turned into ``left(`` and ``right)`` pairs.

**Comment by lervag** [comment]
Interesting request. There is already a related request (#229), although obviously not the same.

At the moment, `tsd` only works in normal mode. If I understand your suggestion correctly, I think you are asking for a visual mode version that will toggle all visually selected delimiters. I first want to mention that this seems rather nontrivial. However, I agree that this might be useful, and I will give it a shot. If you (or anyone) have any suggestions for how to implement this, I am happy to hear it. Pull requests are of course welcome and will be considered!

As you have probably noticed, there are currently a lot of open issues. I hope you understand that this is rather low priority, and so it might take some time.

**Comment by fuzzybear3965** [comment]
It is similar to #229 only in that it, like #258, suggests a improvement on the handling of delimiters.

I don't know Vimscript. It's something I would like to learn and working on vimtex would be a great way to learn it. But, in case I don't get around to doing this myself (because I find myself rather busy these days, too) I have a suggestion for implementing this:

   1. Use a regular expression to find all instances of the selected delimiter, which is denoted in command mode by typing ``:tdX`` where ``X`` is the delimiter to "``\left``-``\right``"-itize. E.g. visually selecting a block and executing the command ``:'<,'>td(`` would search for ``(``s and ``)``s, turning them into ``\left(``s and ``\rights``, respectively.

**Comment by lervag** [comment]
I've pushed a very large refactoring of how delimiters are handled. The toggle will now by default toggle between these pairs of modifiers:

```vim
[
\ ['\left', '\right'],
\ ['\bigl', '\bigr'],
\ ['\Bigl', '\Bigr'],
\ ['\biggl', '\biggr'],
\ ['\Biggl', '\Biggr'],
\ ['\big', '\big'],
\ ['\Big', '\Big'],
\ ['\bigg', '\bigg'],
\ ['\Bigg', '\Bigg'],
\]
```

It can be controlled with `g:vimtex_delim_toggle_mod_list`. It doesn't solve this issue, but it seems very relevant.

**Comment by fuzzybear3965** [comment]
That is an improvement to the handling of delimiters. It seems that the default is actually just to toggle between `left(` and `right)`. This is consistent  with the documentation.

It would still be really awesome if nested delimiters could be toggled in visual mode. That way, swaths of delimiters could be changed with a few keypresses. I don't take the time to write `left(` and `right)`, so I default to using `(` and `)`. I often discover, later, that I missed including a delimiter and, because I'm not using `left(` and `right)` I received no compiler feedback.

**Comment by lervag** [comment]
I've just added the toggle for visual mode. It seems to work as expected on my end. I still need to update the docs.

The current implementation preserves the visual selection, but it does not preserve which "end" you are on, i.e. it always uses the bottom right corner as the cursor position after `tsd`. I'm looking for a way to solve this.

**Comment by lervag** [comment]
Ok, I think this should suffice. Please test, I'll be happy to get some feedback!

**Comment by fuzzybear3965** [comment]
@lervag, this is awesome! This worked really well with some complicated nesting of `(` and `)` and `[` and `]` but it seems to fail with `|` vertical bar/pipe. Consider the following text:

```
\[
   ( a | b | [ c ]  { d } ) .
\]
```
This appears as
```
\[
   \left( a \left| b \left| \left[ c \right]  { d } \right)
\]
```
More simply:
```
\[
| a |
\]
```
becomes
```
\[
\left| a \left|
\]
```
This feature implementation (performed by `vimtex#delim#toggle_modifier(...)`,  I think), also seems to disregard `{` as a delimiter as instances af `{` or `}` in the text are ignored when visually selected and used with `tsd`.

But, for most uses of this feature, it works perfectly. Parentheses are detected in nested order and toggled perfectly. Good job!

**Comment by lervag** [comment]
Ah, yes. I think the problem is that the bar is not really a good delimiter. It is very difficult to parse correctly, since there is not way of knowing whether a bar is to be left or right.

I'll look at it, but in the meantime, I propose that you consider to use `\lvert x \rvert` instead of `| x |`.

**Comment by fuzzybear3965** [comment]
I see. I tried reviewing the source for `vimtex#delim#toggle_modifier(...)` but, unfortunately, I couldn't understand it. It seems, though, that the following implementation for a toggle function would work well:

1. Find the type and positions of all delimiters that have been selected visually.
1. For the first delimiter, find the matching delimiter whose distance from the first delimiter is the greatest.
1. Knowing the positions of the two delimiters, toggle both of them.
1. Repeat the above for all inner delimiters. If this is impossible because delimiters don't match, then a compile-time error will generate and the author should add the matching delimiter 

With the above algorithm and a hash table/set of key-value pairs indicating which left delimiter maps to the right delimiter, you are good to go. You can even handle the case of the `|` delimiters.

**Comment by lervag** [comment]
Nah, actually, IMHO the code is already quite good. It already does most of what you describe, and it uses the existing APIs for parsing delimiters and pairs of delimiters. The thing is, as I mentioned, that it is very difficult to parse `| x |`. I'll look into it, but the problem is more general. You'll notice that it doesn't work for other delimiter commands, e.g. `%`. I've opened a new issue for this (#398).

Regarding the `{d}`: This is not a delimiter. `\{d\}` is.

**Comment by fuzzybear3965** [comment]
Hm, it doesn't seem hard to parse. You find the first `|` and the last `|` and you call those `left` and `right`, respectively; rinse and repeat. Am I misunderstanding something?

You're right `{` and `}` wrap around an object in LaTeX and won't be rendered. That was my mistake. I see that your code works fine with `\{` and `\}`.

**Comment by lervag** [comment]
No, but this assumes that the visual selection is always proper. E.g., in this example, your suggestion might not work:

```tex
\[
  | first part ( x [ y ] ) | = | second part ( x [ y ] ) |
\]
\```

If you select from `first part` to `second` (inclusive), then your method would give `\left| = \right|`.

My point is that it is difficult to parse this in general. Given your assumption that the visual selection is good, one could parse correctly. But I don't want to make that assumption, because I know I personally don't always use "perfect" selections. Thus I think one should either _not_ match the bars, or one should find some more general algorithm to decide if the bar is an opening or a closing bar.

One heuristic/method that I am considering is to weight bars by distance. That is, when I find one bar, then I search both ways for a second bar. If I find two corresponding bars, then I choose the one that is closest. The problem is that I can easily create examples where this also fails.

Could we follow up this discussion in #398?

**Comment by fuzzybear3965** [comment]
I'm leaving my response in #398.
